include/boost/json/detail/impl/number.ipp

```diff
@@ -636,16 +688,31 @@ finish(
         break;
 
     case state::exp3:
-        exp_ += off_;
-        if(eneg_)
+    {
+        if (eneg_)
             exp_ = -exp_;
+
+        if (pos_ == 0)
+        {
+            // abs(mantissa) < 1
+            auto start = dig_ - sig_;
+            auto exponent_adjust = -start - 1;
+            exp_ += exponent_adjust;
```

Comment:
despite all the naming fanfare it produces warnings:
```
2>C:\Users\vinnie\src\boost\libs\json\include\boost/json/detail/impl/number.ipp(705,36): warning C4244: '+=': conversion from 'int' to 'short', possible loss of data
2>C:\Users\vinnie\src\boost\libs\json\include\boost/json/detail/impl/number.ipp(711,36): warning C4244: '+=': conversion from 'int' to 'short', possible loss of data
```

Reply:
When you take away the unnecessary local variables and simplify the formulas, the interesting relationship between the values becomes visible:
```
        if (pos_ == 0)
        {
            // abs(mantissa) < 1
            exp_ = static_cast<short>(
                exp_ + sig_ - dig_ - 1);
        }
        else
        {
            // abs(mantissa) >= 1
            exp_ += static_cast<short>(pos_ - sig_);
        }
```

Reply:
I don't know if my problem is related to your discussion. I have a simple Catch2 test for boost.json 1.75.0 that fails because float/double values are not written/read properly:

```cpp
TEST_CASE("boost::json")
{
    // There seems to be an issue with number formatting.

    // create a boost::json::value
    boost::json::value jv_original = {
        {"pi", 3.14159265359},
        {"e", 2.71828182846}};

    // serialize the boost::json::value to a string
    std::string s{boost::json::serialize(jv_original)};
    // and write the string to a file
    std::string file_name = std::tmpnam(nullptr);
    file_name += ".json";
    std::ofstream ofs(file_name);
    ofs << s;
    ofs.close();

    // now read the file into a string
    std::ifstream ifs(file_name);
    std::string content((std::istreambuf_iterator<char>(ifs)),
                        (std::istreambuf_iterator<char>()));
    ifs.close();
    // and create another boost::json::value from that string
    boost::system::error_code parseErr;
    boost::json::value jv_unserialized = boost::json::parse(content, parseErr);

    // compare both boost::json::value objects
    CHECK(jv_original == jv_unserialized);

} // TEST_CASE("boost::json")
```

The error message from Catch2 is
```bash
[ctest]   CHECK( jv_original == jv_unserialized )
[ctest] with expansion:
[ctest]   {"pi":3.14159265359E0,"e":2.71828182846E0}
[ctest]   ==
[ctest]   {"pi":3.1415926535899996E0,"e":2.71828182846E0}
```

Reply:
Please test against recent Boost version. I can't reprodcue your issue locally on 1.81.0

Reply:
Thanks for your answer, @grisumbras.

You're right. I built and installed Boost 1.81.0 (still running on Ubuntu 20.04) and the failure is gone.

However, I use Boost.JSON 1.81.0 in one of my classes and still have the same problem with this value in a JSON file

```json
"latitude": 79.824228,
```

Putting this value in the code above results in success. But my compare(lhs, rhs) method says:

```bash
[ctest] values of "latitude" differ:
[ctest]   lhs 7.9824228E1
[ctest]   rhs 7.98242E1
```

I'll investigate further ...